### PR TITLE
Update Compose BOM to 2024.02.00. Fixes #1218

### LINKED ISSUE #1218: Update Compose BOM to latest version
Project is currently using Compose BOM 2023.10.01, latest version is 2024.02.00. Source: https://developer.android.com/jetpack/compose/bom/bom-mapping

### PR DESCRIPTION
**Pull request by dturner**
What have I done and why? 
- Updated the [Compose BOM](https://developer.android.com/jetpack/compose/bom/bom-mapping) from 2023.10.01 to 2024.02.00 (Compose 1.6.2) 
- Updated deprecated/changed Material3 components in line with recommended guidance
- Fixed a couple of minor build warnings
- Replaced `BoxWithConstraints` with `Box`
- Changed font heights, alignment and spacing so that the current UI layouts stay the same. In Compose 1.6 (introduced by this BOM) `includeFontPadding` defaults to `false`. This causes some unwanted layout changes. To preserve the current UI I had to tweak the typography styles using `LineHeightStyle` and make some slight layout changes. The new layouts aren't a pixel perfect match (as can be seen in the screenshot diffs) but they are close enough so that no-one should notice.  

**Notes on screenshot diffing**
Updating the layouts to match previous screenshots was a time consuming process. In case anyone else needs to do it, here's what I did: 

- Clone the repo, checkout the commit/branch which contains the "old" screenshots i.e. the ones which are considered to be correct 
- Record new screenshot tests`./gradlew recordRoborazziDemoDebug --rerun-tests`. Note the `--rerun-tests` ensures that new screenshots are actually recorded, without it I would sometimes find that they weren't updated despite code changes
- Copy the screenshot(s) folders to "screenshots-old" in the same parent folder - this is to move them out of source control
- Checkout your current branch (the one which will contain the new screenshots), make UI changes and record new screenshots
- Use [Beyond Compare 4](https://www.scootersoftware.com/) to do a folder comparison between the old and new screenshots folders. It has some extremely useful features like being able to offset one image on the other and do a pixel by pixel comparison with surrounding pixels: 

image

### CODE REVIEW

**Comment by JoseAlcerreca** [comment]
Looks like if you re-run a job after the screenshots were generated, it fails when trying to commit the new ones. Can you add another commit or force-push without the 9e64 commit?

**Comment by LinX64** [comment]
> Looks like if you re-run a job after the screenshots were generated, it fails when trying to commit the new ones. Can you add another commit or force-push without the 9e64 commit?

As a note, a **rebase** is required to apply the changes from your previous [merged PR](https://github.com/android/nowinandroid/pull/1219) about CI to solve the emulator issue, please see [here](https://github.com/android/nowinandroid/actions/runs/7963490467/job/21740038239#step:8:82).

File: feature/search/src/test/kotlin/com/google/samples/apps/nowinandroid/feature/search/SearchViewModelTest.kt [hunk 1]
```diff
@@ -99,7 +99,7 @@ class SearchViewModelTest {
         searchContentsRepository.addNewsResources(newsResourcesTestData)
         searchContentsRepository.addTopics(topicsTestData)
 
-        val result = viewModel.searchResultUiState.value
+        // val result = viewModel.searchResultUiState.value
```
Comment (SimonMarquis) [inline feature/search/src/test/kotlin/com/google/samples/apps/nowinandroid/feature/search/SearchViewModelTest.kt:102]: btw, the fix is here and waiting for being merged:
- #1040

File: app-nia-catalog/dependencies/releaseRuntimeClasspath.txt [hunk 2]
```diff
@@ -1,56 +1,58 @@
...
+androidx.compose.ui:ui-util:1.6.1
+androidx.compose.ui:ui:1.6.1
+androidx.compose:compose-bom:2024.02.00
 androidx.concurrent:concurrent-futures:1.1.0
 androidx.core:core-ktx:1.12.0
 androidx.core:core:1.12.0
 androidx.customview:customview-poolingcontainer:1.0.0
 androidx.customview:customview:1.0.0
-androidx.emoji2:emoji2:1.4.0
+androidx.emoji2:emoji2:1.3.0
```
Comment (dturner) [inline app-nia-catalog/dependencies/releaseRuntimeClasspath.txt:55]: Which dependency caused this downgrade?

File: app/dependencies/prodReleaseRuntimeClasspath.txt [hunk 3]
```diff
@@ -61,8 +63,8 @@ androidx.datastore:datastore-preferences:1.0.0
 androidx.datastore:datastore:1.0.0
 androidx.documentfile:documentfile:1.0.0
 androidx.drawerlayout:drawerlayout:1.0.0
-androidx.emoji2:emoji2-views-helper:1.4.0
-androidx.emoji2:emoji2:1.4.0
+androidx.emoji2:emoji2-views-helper:1.3.0
```
Comment (dturner) [inline app/dependencies/prodReleaseRuntimeClasspath.txt:66]: More downgrades

File: app-nia-catalog/dependencies/releaseRuntimeClasspath.txt (hunk 2 above)
Comment (dturner) [reply app-nia-catalog/dependencies/releaseRuntimeClasspath.txt:55]: androidx.compose.ui:ui-text-android:1.6.1 has a transitive dependency on androidx.emoji2.emoji2.  The upper bound of versions supported changed from 1.4.0 to 1.3.0. 

Dependency information obtained using `./gradlew :app:dependencies > deps.txt`

File: app/dependencies/prodReleaseRuntimeClasspath.txt (hunk 3 above)
Comment (dturner) [reply app/dependencies/prodReleaseRuntimeClasspath.txt:66]: Same reason as previous comment.

**Review by alexvanyo — APPROVED** [review]
Thank you for the detailed description for the text style changes! Resulting screenshot changes look good, either barely noticable or strictly an improvement (like the navigation bar one at large font scales)

**Pull request merged**